Blue typescript

skqksh opened the pull request:
## Proposed changes

- Support Typescript (#157)

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/caver-js/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/caver-js)
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

-#157

## Further comments

I added declare files to support typescript. I focused to change files where are in official docs
Lots of testcases include unnecessary options(parameters, etc) for the test, so I hope to change *.js files to *.ts
I believe that It will help to find bugs or prevent other type errors

skqksh commented:
@jimni1222 

Awesome ~! Thanks for your help

![boo](https://user-images.githubusercontent.com/16040418/98619483-af42b700-2346-11eb-9401-a7646a94d266.jpg)

jimni1222 commented:
@skqksh 

Yeahhh~ now this PR has passed the circle ci test. 👍 

Can i ask you something?
What is the criteria for selecting files converted to ts?

And as you can see in the [common architecture](http://kips.klaytn.com/KIPs/kip-34), Caver's interface naming rule is ʻIXXX`. Please change the interface name from `XXX_I` to ʻIXXX`.

skqksh commented:
@jimni1222 

I changed interface's name

- What is the criteria for selecting files converted to ts?
I converted files written in Klaytn-Docs, first. Some codes seem not using anymore
I didn't convert some files which are still alive in caver-js. I just remained that for someone who want to try(Maybe me of the future?)
I hope you guys to change other testcases to .ts, It must be good to make nicer API

jimni1222 commented on packages/caver-account/src/accountKey/accountKeyPublic.d.ts:
```diff
@@ -0,0 +1,86 @@
+ * Representing an AccountKeyPublic.
+ * @class
+ */
+export interface IAccountKeyPublic {
```
Why should interface and class be defined in d.ts file at the same time? What purpose is each used for?

skqksh replied:
@jimni1222 
yes. I made it.

To make instance of class in as a class property, I had to define interface
You can refer to [Typescript Docs](https://www.typescriptlang.org/docs/handbook/interfaces.html#class-types)
For example, if you set type as class(not interface), it returns error
```
export class Caver {
    ...
    account: Account // not account: IAccount
    ...
}
```
`
This expression is not constructable.
  Type 'Account' has no construct signatures.
`

jimni1222 commented:
@skqksh Please add the license text at the top of each newly created file. One blank line is required after the license text.

jimni1222 commented on index.d.ts:
```diff
@@ -0,0 +1,71 @@
+
+    constructor(provider: any, net?: any)
+
+    account: IAccount
```
Why are some types defined as interfaces? By what criteria is it divided?

skqksh replied:
@jimni1222 
the reason is same with above

jimni1222 commented on tsconfig.json:
```diff
@@ -0,0 +1,70 @@
+    // "noFallthroughCasesInSwitch": true,    /* Report errors for fallthrough cases in switch statement. */
+
+    /* Module Resolution Options */
+     "moduleResolution": "node",            /* Specify module resolution strategy: 'node' (Node.js) or 'classic' (TypeScript pre-1.6). */
```
```suggestion
    "moduleResolution": "node",            /* Specify module resolution strategy: 'node' (Node.js) or 'classic' (TypeScript pre-1.6). */
```

jimni1222 commented on test/packages/caver.account.ts:
```diff
@@ -1,50 +1,46 @@
-/*
-    Copyright 2020 The caver-js Authors
-    This file is part of the caver-js library.
 
-    The caver-js library is free software: you can redistribute it and/or modify
-    it under the terms of the GNU Lesser General Public License as published by
-    the Free Software Foundation, either version 3 of the License, or
-    (at your option) any later version.
-
-    The caver-js library is distributed in the hope that it will be useful,
-    but WITHOUT ANY WARRANTY; without even the implied warranty of
-    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
-    GNU Lesser General Public License for more details.
-
-    You should have received a copy of the GNU Lesser General Public License
-    along with the caver-js. If not, see <http://www.gnu.org/licenses/>.
-*/
```
Please don't remove license comments

skqksh replied:
@jimni1222 

okay~

skqksh commented:
> @skqksh Please add the license text at the top of each newly created file. One blank line is required after the license text.

okay~

devhyunjae commented:
@here Any progress in here? Need ts support!

jimni1222 commented:
@devhyunjae Sorry, but this PR is still under review.
@skqksh Please resolve the conflict.( and please rebase on latest dev branch :) ) And when will the request made in the comment above be reflected?

skqksh commented:
@jimni1222 
I found that some of my codes(usage of  typescript) are wrong.
I'll rework it someday...
But now, Please close this PR. T^T

jimni1222 commented:
@skqksh Oh... I see! First, I will close this PR then. :) Thank you for always.
